lib/state/domain/notes.ts

```diff
@@ -1,6 +1,30 @@
 import { noteBucket } from './buckets';
 import isEmailTag from '../../utils/is-email-tag';
 import { createTag } from './tags';
+import * as T from '../../types';
+
+export const toggleSystemTag = (
+  note: T.NoteEntity,
+  systemTag: T.SystemTag,
+  shouldHaveTag: boolean
+) => {
+  const {
+    data: { systemTags = [] },
+  } = note;
+  const hasTagAlready = systemTags.includes(systemTag);
+
+  return hasTagAlready !== shouldHaveTag
+    ? {
+        ...note,
+        data: {
+          ...note.data,
```

Comment:
What happens when the note data is updating as this is happening?

Reply:
This is why the bug reported here as `1` is occurring: https://github.com/Automattic/simplenote-electron/issues/1911#issue-566462994

Reply:
this function should only adjust the system tag and nothing else.

at least by moving it here as a side-effect we eliminate one part of the data race which was dispatch-update-dispatch before. I don't think this is the place where we can resolve that bug either.

note that before we sent the update to simperium and then dispatched `selectNote`. in this version we just ship off the update to simperium and wait for a remote update to trigger the re-render cycle.

I guess we should be watching for these actions in the `note` reducer too to make sure it updates instantly.

can you elaborate more on the conditions you see "when the note data is updating as this is happening"? Are you talking about updates in the editor or remote updates?

we should have a guarantee that no editor updates are happening because of the single-threaded nature of JavaScript. when we click on "pin note" then the action will dispatch and all reducers will run _before_ interactivity returns to the UI.

I think that leaves remote updates and if that's the case then I think the middleware will still have to run before `node-simperium` processes the incoming message and therefore that means this change should be translated when applying the patch.